crates/bevy_ecs/src/component.rs

```diff
@@ -952,6 +1000,283 @@ impl Components {
+        };
+
+        for (component_id, component) in recursive_requires.iter().cloned() {
+            // SAFETY: The component is in the list of required components, so it must exist already.
+            let required_components = unsafe {
+                self.get_required_components_mut(requiree)
+                    .debug_checked_unwrap()
+            };
+
+            // Register the required component for the requiree.
+            // SAFETY: Component ID and constructor match the ones on the original requiree.
+            //         The original requiree is responsible for making sure the registration is safe.
+            unsafe {
+                required_components.register_dynamic(
+                    component_id,
+                    component.constructor,
+                    component.inheritance_depth,
+                );
+            };
+
+            // Add the requiree to the list of components that require the required component.
+            // SAFETY: The component is in the list of required components, so it must exist already.
+            let required_by = unsafe {
+                self.get_required_by_mut(component_id)
+                    .debug_checked_unwrap()
+            };
+            required_by.insert(requiree);
+        }
+
+        // Propagate the new required components up the chain to all components that require the requiree.
+
+        if let Some(required_by) = self.get_required_by(requiree).cloned() {
+            for &required_by_id in required_by.iter() {
+                // SAFETY: The component is in the list of required components, so it must exist already.
+                let required_components = unsafe {
+                    self.get_required_components_mut(required_by_id)
+                        .debug_checked_unwrap()
+                };
+
+                // Register the original required component for the requiree.
+                // The inheritance depth is `1` since this is a component required by the original requiree.
+                required_components.register_by_id(required, constructor, 1);
+
+                for (component_id, component) in recursive_requires.iter() {
+                    // Register the required component.
+                    // The inheritance depth is increased by `1` since this is a component required by the original required component.
+                    // SAFETY: Component ID and constructor match the ones on the original requiree.
+                    //         The original requiree is responsible for making sure the registration is safe.
+                    unsafe {
+                        required_components.register_dynamic(
+                            *component_id,
+                            component.constructor.clone(),
+                            component.inheritance_depth + 1,
+                        );
+                    };
+                }
+            }
+        }
```

Comment:
Yeah, agreed. I extracted at least the recursive registration logic into their own methods now. It could maybe be split even further, but I think it should be reasonably easy to follow now